Add support for the free-threaded build

ngoldbaum opened the pull request:
Fixes #126.

Adds support for the free-threaded build of Python 3.13 and 3.14 by improving the thread safety of CFFI internals.

## Overview

Mutable state that is accessible to more than one thread in the CFFI backend is now mediated via atomic operations, critical sections, or mutexes, depending on the use-case:

- Replace use of PyDict and PyWeakref C APIs producing borrowed references with new APIs (or shims on older versions) that produce strong references.
- `get_primitive_type`: We initialize all primitive types at module initialization, instead of doing them lazily. This is pretty inexpensive (~100 µs).
- `file_struct`: now initialized during module initialization
- `_get_ct_int`: now initialized during module initialization
- `init_once_cache`: The cache itself is initialized under a critical section and we use APIs that return strong references instead of borrowed references.
- `malloc_closure`: use a global mutex
- Moved mutable state to uint8 flags synchronized via atomic operations.
- `b_complete_struct_or_union`: Added critical section.
- Make the test suite runnable under [`pytest-run-parallel`](https://github.com/Quansight-Labs/pytest-run-parallel) and sets up `pytest-run-parallel` Linux, Mac, and Windows CI as well as a run using a TSAN-instrumented build running on Python 3.14.

## Open questions

- More docs?
- The `ctypes` module in CPython isn't thread-safe in 3.13t. Should we worry about the `ctypes` backend on 3.13t?
- More explicit multithreaded tests?

cc @kumaraditya303 @colesbury

@colesbury also wants to do another round of code review on the final version of this PR.

mattip commented:
> @colesbury also wants to do another round of code review on the final version of this PR.

Is this still relevant or is this ready for whatever we can do to review it?

ngoldbaum commented:
I think it's ready for anyone to review who wants to. 

The last I heard from Sam was that he thought things overall looked good but might have comments. I have a meeting with the Meta team tomorrow where I'll bring this PR up.

mattip commented:
I think this is a big enough change that it should be in a separate release. We should release 1.18 as soon as possible, and create a new 1.19 (or 2.0?) release that supports free threading.

kumaraditya303 commented on src/c/_cffi_backend.c:
```diff
@@ -6716,7 +6884,11 @@ static CTypeDescrObject *direct_typeoffsetof(CTypeDescrObject *ct,
             return NULL;
         }
+#ifdef Py_GIL_DISABLED
+        cf = (CFieldObject *)PyDict_GetItem((PyObject *)cffi_atomic_load((void **)&ct->ct_stuff), fieldname);
+#else
```
I tried changing the load to be non atomic after `force_lazy_struct`, it leads to data races you can see here https://github.com/ngoldbaum/cffi-ft/actions/runs/15906594432/job/44863599643?pr=1#step:5:1657

I think that even if the field is initialized, since the store is done atomically tsan would complain if it is loaded without atomic, I am not sure.

ngoldbaum replied:
This was fixed by adding a new global lock and atomic field for a fast path.

mattip commented on src/c/ffi_obj.c:
```diff
@@ -185,16 +187,35 @@ static CTypeDescrObject *_ffi_type(FFIObject *ffi, PyObject *arg,
+            struct _cffi_parse_info_s info;
+            info.ctx = &ffi->types_builder.ctx;
+            info.output_size = FFI_COMPLEXITY_OUTPUT;
+#ifdef Py_GIL_DISABLED
+            info.output = PyMem_Malloc(FFI_COMPLEXITY_OUTPUT * sizeof(_cffi_opcode_t));
+            if (info.output == NULL) {
+                PyErr_NoMemory();
+                return NULL;
+            }
+#else
```
Makes sense to me. Since the result is cached, this will be hit relatively rarely.

ngoldbaum commented:
@mattip @nitzmahone I think this is ready from our end - @kumaraditya303 and @colesbury figured out the thread safety issue we ran into last week. I also resolved the remaining comments from the last round of review.

ngoldbaum commented on testing/cffi1/test_ffi_obj.py:
```diff
@@ -32,7 +32,11 @@ def test_ffi_cache_type():
 def test_ffi_type_not_immortal():
+    if hasattr(sys, '_is_gil_enabled') and not sys._is_gil_enabled():
+        pytest.skip("fails on free-threaded builds")
+
```
This is failing on the GIL-enabled build too - https://github.com/ngoldbaum/cffi-ft/actions/runs/16023540072/job/45205948721

I think this change is incorrect and something else is causing this to fail. @kumaraditya303 are looking at it. Again, sorry for the noise...

kumaraditya303 commented on src/c/_cffi_backend.c:
```diff
@@ -225,22 +269,32 @@
+/* flags that are mutable at runtime, stored separately in ct_flags_mut to avoid
+   races with ct_flags
+
+   these are read and set without using atomic operations because a lock is held
```
```suggestion
   these are read and set without using atomic operations because CFFI_LOCK is held
```

kumaraditya303 commented on src/c/realize_c_type.c:
```diff
@@ -49,12 +57,40 @@ static int init_global_types_dict(PyObject *ffi_type_dict)
+        return err;
+
+#ifdef Py_GIL_DISABLED
+    // Ensure that all primitive types are intitalized to avoid race conditions
```
```suggestion
    // Ensure that all primitive types are initialised to avoid race conditions
```

ngoldbaum commented:
> Would it be too much to ask that you try out a few prime CFFI libraries with good test suites, like cryptography and curl_cffi to see if their tests pass? I don't know how hard it would be to switch their build systems to use this PR.

Sure, I can try those two. We already tried argon2-cffi with an older version of this PR and saw no races running its test suite in parallel.

ngoldbaum commented:
> like cryptography

Take a look at this CI run on cryptography: https://github.com/pyca/cryptography/actions/runs/16029864342/job/45227459492

All the failures are because of the very hacky way I added this PR as a depenency to cryptography, which breaks the test runs that depend on creating a lockfile.

> and curl_cffi

This is trickier - right now it's using a cibuildwheel version from March 2024 and updating the CI config looks nontrivial. I'd also need to figure out how to get libffi on the build container to allow cffi to build on the CI host---cryptography already has libffi set up so it's not an issue there.

I'm also not able to run the the tests on a local build using a GIL-enabled 3.12 interpreter - not sure what I'm doing wrong.

> argon2-cffi 

I tried this again with the latest version of the PR. The tests all pass on the free-threaded build and the GIL-enabled build. The tests also pass if I force the GIL to be disabled. There's an explicitly multithreaded test in the argon2-cffi tests that I added earlier this year.

ngoldbaum commented:
> Cool. So if I am reading that correctly, the new CFFI works everywhere, even on non-free-threading builds. The Python3.8 failures are not connected.

yup!

mattip reviewed (commented):
From what I can gather, this seems to make sense. There might be some design choices around granularity of the locks that could impact performance on the free-threaded build, but I think those can be adjusted in subsequent PRs to address any performance impact from lock churning vs. thread suspension without changing user-facing API.

As discussed, this will be part of a 2.0 release, so the version needs updating, not necessarily in this PR.

ngoldbaum commented:
@mattip @nitzmahone @arigo gentle ping on this. Is there anything more that needs to be done here? I can deal with any comments or concerns you have if you're waiting for time to do a full once-over on this.

mattip reviewed (approved):
Looks reasonable to me, once CI passes

mattip commented:
CI is passing

nitzmahone commented:
Spent a good chunk of today trying to wrap up my promised review- in general I think we're good to go, but there were a couple little things:
1) There are still several instances where the GIL is getting re-enabled when we load a test extension- in at least one case I tried locally *during collection*, so basically the entire test run was getting run with the GIL enabled. I see someone added a conftest hook to sample that before and after the test session and report a mismatch (yay!), but we're not actually enforcing it for the free-threaded test targets. I've added a conditional "used by everything" fixture locally that asserts the GIL state before and after each test when we should be GIL-free- working my way through fixing the remaining tests that were causing it to be re-enabled. Hopefully that won't expose any new TSan issues...
2) Looking at several of the cases from 1), it feels like there needs to be an extension kwarg (or some other knob) to force `Py_GIL_DISABLED` to be defined (or explicitly undefined) for a given generated extension, rather than relying solely on whatever preprocessor state happens to be floating around when we build an extension.

I'll be back at it later tonight or tomorrow morning- hopefully can get those bashed out and assuming no new issues from a full CI run, get a 2.0.0b1 shipped tomorrow.

ngoldbaum commented:
OK great, please feel free to push directly to this PR branch if that's convenient for you. I'll also try to reproduce what you were seeing with the test extensions.

Having a per-test check in pytest-run-parallel like you describe is probably a good idea. We could add it to the test executor, which is already wrapped in a try/finally.

Of course that doesn't help if the GIL gets re-enabled in a subprocess, which might be what's happening here.

ngoldbaum commented:
I looked at the extensions created by the tests and it looks like they all set `Py_MOD_GIL_NOT_USED`, at least on my Mac testing environment. I'm also unable to reproduce the GIL being re-enabled. It looks like it's not happening on any of the CI runs either, unless I'm missing something.

I spent a little time drilling down into the extensions generated by `recompiler.py`, but ultimately all the modules created there are modified versions of the "main" `_cffi_backend` module, which has `Py_MOD_GIL_NOT_USED` set.

@colesbury tells me he thinks the most likely result of not having `Py_GIL_DISABLED` set is a crash, rather than the GIL being re-enabled.

I'll wait until I have more info to proceed further.

colesbury commented:
I'm not sure why @nitzmahone is seeing GIL enabled warnings. I don't see any locally and I don't see any in the GitHub CI logs. Could you have some unrelated or stale modules in the same directory that are pulled in by `pytest`?

The Windows explanation doesn't make sense to me. First, `setuptools` passed the necessary `Py_GIL_DISABLED` compiler flag to work around the Windows installer issue that @ngoldbaum mentioned. Even if that didn't happen, a missing `Py_GIL_DISABLED` flag when compiling will lead to crashes on import due to incorrect struct definitions; I don't think you'd see GIL enabled warnings.

@nitzmahone - when you get a chance to look at it more, would you please share the relevant logs?

nitzmahone commented:
The issues I've been seeing were all from manual poking around on my Linux dev box (Fedora 42 x86_64) with a bone-stock build from `pyenv install 3.14.0rc1t`.

*Most* of the test extensions are being correctly built with `Py_GIL_DISABLED`, which was why I hacked in the more granular per-test pre/post check (which currently has to be run sans worker threads to blame the correct test). I'd just started digging into the why last night when I ran out of time. From my past experience with it, CFFI's test suite is notoriously "leaky", and there are a number of one-off tests sprinkled around that do things ... differently. It's also complicated by so many operational modes- I've fixed some tests in the past that were directly (or indirectly) invoking the wrong Python and/or config, so it's possible that's what's happening here. I was just disabling the individual tests that were re-enabling the GIL last night to come up with the full list, so I haven't dived into the why yet.

Now that I'm less pudding-brain, I'll go back over it with fresh eyes and make sure I'm not getting tripped up by random stale non-`t` test build artifacts or something else. I might also temporarily switch to an xdist/forked run model like we do for Ansible- getting rid of intra-test interpreter state leakage makes it a lot easier to pinpoint problems.

More to come...

nitzmahone commented:
Hrm, after kicking off a few manual workflow runs with all targets enabled to simulate a release, the [Linux](https://github.com/nitzmahone/cffi/actions/runs/16509850542/job/46689417654) and [MacOS](https://github.com/nitzmahone/cffi/actions/runs/16509850542/job/46689417658) 3.13t parallel runs (skipped for PRs) are both segfaulting in CI in exactly the same place (`testing/cffi1/test_pkgconfig.py`) where 3.14t is fine.

I'm able to very reliably repro the segfault locally on 3.13.5t with:
```
pytest testing/cffi1/test_re_python.py --parallel-threads=3  # 2 breaks sometimes, >=3 breaks quite reliably
```
where various similar permutations seem fine on 3.14.0rc1t.

I'm not opposed to proceeding with merge/beta, but I would really like to know what's going on there. I need to hang it up for the day pretty soon, and I'm unavailable on Friday morning. If someone else that can repro this locally wants to try and catch it in the act against a debug build, cool, otherwise I'll take a stab at it Friday afternoon.

ngoldbaum commented:
Argh! So close...

I didn't see this when I triggered the full CI back in June: https://github.com/ngoldbaum/cffi-ft/actions/runs/15769464108. I Just double-checked and that run *did* include the commit that makes sure the GIL doesn't get re-enabled in the tests, so that test run did pass on 3.13t.

Too bad I didn't think to redo that exercise in the month since then. Apologies for the oversight.

I ran the tests tonight using 3.13.5t on my Mac and I see similar results ([TSAN reports and random test failures](https://gist.github.com/ngoldbaum/40eb8dfcd7c14190a61d50bb2b074295), although no segfaults).

We did some updates to the PR after I triggered that full CI run - in particular we adjusted the locking strategy to do something a lot simpler - just use a single critical section on a single globally allocated dummy PyObject. On Python 3.14, this acts more-or-less like a sort of "per-library GIL". On Python 3.13 there are different semantics for when critical sections can be suspended and I suspect that's the source of the differences here. That's just my guess anyway.

I pinged @kumaraditya303 to help track this down. Hopefully he has time to look at this tomorrow and there is something straightforward we can do, but it may take a little time to work out the correct fix.

Just a thought - maybe we can disable pytest-run-parallel CI for the wheel builds on 3.13t while we work out these issues, just to unblock people who want to enable builds downstream. We can note in the release notes that there are known thread safety issues on 3.13t.

nitzmahone commented:
Yeah, I'd prefer to have the beta working without caveat- hopefully there's a quick fix once we figure out exactly what's going on. Painfully slow as it is, it'd probably be wise to at least temporarily enable the full test suite for a Windows target or two as well- IIRC I had the whole thing passing for either early 3.14 alpha or late 3.13 beta (only tested with GIL-ful threading), so not sure what its current operational state is. I'll pick that up Friday afternoon as well.

ngoldbaum commented:
@kumaraditya303 did some digging and we're now pretty confident the issue is what I described above: recursive critical sections can be suspended more often in 3.13 than they could be in 3.14. This comes down to this change: https://github.com/python/cpython/pull/128126. While that is billed as a performance improvement, it also has a side-effect of making recursive critical sections behave more like recursive locks in 3.14 than they did in 3.13.

We now think the clearest path forward is probably just to not try to support 3.13t in CFFI, at least not at first. If it turns out that 3.13t support is critical for whatever reason, we can add support later, but it'd be a shame to not ship 3.14t support using our current approach when we know that works fine.

Another reason not to support 3.13t is there are also races in CPython internals that are triggered in the CFFI test suite (see [the gist](https://gist.github.com/ngoldbaum/40eb8dfcd7c14190a61d50bb2b074295) I shared yesterday). Some of them are coming from PyDict internals -- CFFI using python dictionaries heavily in its implementation -- and as I understand it the fixes for these bugs probably won't be backported to 3.13.

If we do need to support 3.13t then we probably need to replace the `CFFI_LOCK` critical section with a recursive mutex. The problem there is there isn't an obvious, portable choice to use in C. In NumPy we solved a similar issue by using C++ standard library features but we probably can't do that here. The Python C API doesn't expose a recursive mutex.

If anyone really does need to use 3.13t, they can, they'll just need to deal with the thread safety issues. You can work around them by initializing types in the main thread before spawning worker threads. Obviously that's not good and CFFI shouldn't officially support that, but 3.13t is experimental as-is and anyone using it can be expected to go a little out of their way to get things working.

@nitzmahone @mattip if you're ok with not supporting 3.13t then I'll go ahead and remove the 3.13t CI and wheel builds from this PR.

ofek commented:
As a user, we are okay with only 3.14 support.

ngoldbaum commented:
See latest commits. I also kicked off a "full" CI run on my fork: https://github.com/ngoldbaum/cffi-ft/actions/runs/16530375859

ngoldbaum commented:
It looks like everything is passing. Still waiting on PPC64le and s390x but that will take a while since they're emulated. Maybe I should re-enable Windows too?

nitzmahone commented:
Yeah, that all looks great- I'm ready to merge if you are.

ngoldbaum commented:
:shipit:

ngoldbaum commented:
Also if you need any support in any tasks for shipping the beta or final release, feel free to ping me 😀

nitzmahone commented:
Nah, just woke up to lots of other things on fire this morning- merging now...

alex commented:
Woooo! Huge thank you to everyone who made this happen!
